File: client/src/types/resume.ts
export type ResumeStatus = 'uploaded' | 'extracting' | 'embedding' | 'completed' | 'failed';

export interface Resume {
  id: string;
  userId: string;
  filename: string;
  filePath: string;
  fileType: 'pdf' | 'image';
  status: ResumeStatus;
  extraction_method?: string;
  rawText?: string;
  createdAt: string;
  updatedAt: string;
}

export interface ResumeSection {
  id: string;
  section_label: string;
  content: string;
  section_order: number;
  metadata?: Record<string, any>;
}

export interface UploadResponse {
  id: string;
  filename: string;
  file_type: string;
  status: string;
  extraction_method?: string;
  created_at: string;
  message: string;
  resume: Resume; // Add this since hooks expect response.data.resume
}

export interface ContactInfo {
  email?: string;
  phone?: string;
  links?: string[];
  location?: string;
}

export interface ExperienceEntry {
  company?: string;
  role?: string;
  dates?: string;
  description?: string;
}

export interface EducationEntry {
  institution?: string;
  degree?: string;
  dates?: string;
  metrics?: string;
}

export interface ProjectEntry {
  name?: string;
  description?: string;
  technologies?: string[];
}

File: server/app/schemas/resume.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime
from uuid import UUID


# ── Upload ─────────────────────────────────────────────────────────────────────

class ResumeUploadResponse(BaseModel):
    id: UUID
    filename: str
    file_type: str
    status: str
    extraction_method: Optional[str] = None
    created_at: datetime
    message: str = "Resume uploaded successfully"

    class Config:
        from_attributes = True


# ── Resume Metadata ────────────────────────────────────────────────────────────

class ResumeResponse(BaseModel):
    id: UUID
    user_id: UUID
    filename: str
    file_path: str
    file_type: str
    status: str
    extraction_method: Optional[str] = None
    raw_text: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ResumeListResponse(BaseModel):
    resumes: List[ResumeResponse]
    total: int


# ── Section ────────────────────────────────────────────────────────────────────

class ResumeSectionResponse(BaseModel):
    id: UUID
    section_label: str
    content: str
    section_order: int
    meta: Optional[Dict[str, Any]] = None

    class Config:
        from_attributes = True


class ResumeSectionsResponse(BaseModel):
    resume_id: UUID
    extraction_method: Optional[str] = None
    sections: List[ResumeSectionResponse]


class ResumeFullJSON(BaseModel):
    """Assembled JSON payload for LLM consumption — from structured store only."""
    doc_id: UUID
    sections: Dict[str, str]


# ── Standardized Extracted Resume ──────────────────────────────────────────────

class ExperienceEntry(BaseModel):
    company: Optional[str] = None
    role: Optional[str] = None
    dates: Optional[str] = None
    description: Optional[str] = None

class EducationEntry(BaseModel):
    institution: Optional[str] = None
    degree: Optional[str] = None
    dates: Optional[str] = None
    metrics: Optional[str] = None

class ProjectEntry(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    technologies: List[str] = Field(default_factory=list)

class ContactInfo(BaseModel):
    email: Optional[str] = None
    phone: Optional[str] = None
    links: List[str] = Field(default_factory=list)
    location: Optional[str] = None

class StandardizedResume(BaseModel):
    """The final strict JSON representation of a resume."""
    name: Optional[str] = None
    role: Optional[str] = None
    contact_info: ContactInfo = Field(default_factory=ContactInfo)
    professional_summary: Optional[str] = None
    skills: List[str] = Field(default_factory=list)
    experience: List[ExperienceEntry] = Field(default_factory=list)
    education: List[EducationEntry] = Field(default_factory=list)
    projects: List[ProjectEntry] = Field(default_factory=list)
    certifications: List[str] = Field(default_factory=list)
    publications: List[str] = Field(default_factory=list)


# ── Retrieval (PDF path only) ──────────────────────────────────────────────────

class RetrievalRequest(BaseModel):
    resume_id: UUID
    query: str
    section_filter: Optional[str] = None
    top_k: int = Field(default=5, ge=1, le=20)


class RetrievalResult(BaseModel):
    chunk_text: str
    section: str
    score: float
    chunk_index: int


class RetrievalResponse(BaseModel):
    resume_id: UUID
    query: str
    results: List[RetrievalResult]
